[JSC] Add scavenger hook for scavenging external tasks

Achierius opened the pull request:
#### c7d6142e894fc9ca22be5151076c5c3e40e99f77
<pre>
[JSC] Add scavenger hook for scavenging external tasks
<a href="https://bugs.webkit.org/show_bug.cgi?id=287984">https://bugs.webkit.org/show_bug.cgi?id=287984</a>
<a href="https://rdar.apple.com/145154078">rdar://145154078</a>

Reviewed by Yusuke Suzuki.

This patch should strike a balance between convenience and performance.
Using a fixed array of callbacks is suitable because we always
know which systems are going to hook into libpas -- it&apos;s not like
someone will decide at runtime that they need to fall back on
the scavenger.

Re.: the additional mutex, it&apos;s possible to do without it -- we would
just need two atomic integers, &quot;next_open_slot&quot; and
&quot;last_initialized_slot&quot;. However, this is a nontrivial amount of
complexity and much easier to mess up.

* Source/WTF/wtf/SequesteredAllocator.h:
* Source/WTF/wtf/SequesteredImmortalHeap.h:
(WTF::SequesteredImmortalHeap::installScavengerHook):
(WTF::SequesteredImmortalHeap::scavenge):
* Source/WTF/wtf/Threading.cpp:
(WTF::initialize):
* Source/bmalloc/libpas/libpas.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/src/libpas/pas_scavenger.c:
(ensure_data_instance):
(scavenger_thread_main):
(pas_scavenger_try_install_foreign_work_callback):
* Source/bmalloc/libpas/src/libpas/pas_scavenger.h:
* Source/bmalloc/libpas/src/test/ScavengerExternalWorkTests.cpp: Added.
(incrementCounter):
(testCallbacksAreCalledWhenExpected):
(addScavengerExternalWorkTests):
* Source/bmalloc/libpas/src/test/TestHarness.cpp:

Canonical link: <a href="https://commits.webkit.org/290978@main">https://commits.webkit.org/290978@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43a9b874c8f5a16db2033f7c3946445211f23355

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91577 "21 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11108 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/642 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96588 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42259 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11485 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19538 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/70355 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27839 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94578 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8773 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82968 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50681 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8537 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/557 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41430 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/84376 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78856 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/560 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98551 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/90325 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18736 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79357 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18990 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78806 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78583 "Found 1 new API test failure: /WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23080 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11862 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14521 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18730 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112909 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18440 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32730 "Found 16 new JSC stress test failures: microbenchmarks/memcpy-wasm-small.js.dfg-eager, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager-jettison, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-slow-memory, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.default-wasm, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-collect-continuously, wasm.yaml/wasm/stress/repro_1289.js.default-wasm, wasm.yaml/wasm/stress/repro_1289.js.wasm-eager, wasm.yaml/wasm/stress/repro_1289.js.wasm-eager-jettison, wasm.yaml/wasm/stress/repro_1289.js.wasm-slow-memory, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate-catch.js.wasm-eager-jettison ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21901 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20207 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/bmalloc/libpas/src/test/ScavengerExternalWorkTests.cpp:
```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2021 Apple Inc. All rights reserved.
```
2025

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_scavenger.c:
```diff
@@ -387,6 +400,35 @@ static void* scavenger_thread_main(void* arg)
+
+    pthread_mutex_unlock(&data->foreign_work.lock);
+
+    return true;
```
what if the scavenger is suspended at this time? You may need to kick it, no?

Achierius replied:
I leave this up to the caller, since it has to happen multiple times regardless -- like how we already call it whenever we call e.g. pas_try_allocate_common_impl_slow

In my particular case it'll land with the followup patch that hooks the sequestered arena allocator into this

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_scavenger.c:
```diff
@@ -387,6 +400,35 @@ static void* scavenger_thread_main(void* arg)
+
+    data->foreign_work.descriptors[slot].period_log2_ticks = pas_log2(requested_ticks);
+    data->foreign_work.descriptors[slot].func = callback;
+    data->foreign_work.descriptors[slot].userdata = userdata;
```
Emit store-barrier here. So scavenger can see the fully stored slot when `next_open_descriptor` includes this slot.

Achierius committed:
[JSC] Add scavenger hook for scavenging external tasks
https://bugs.webkit.org/show_bug.cgi?id=287984
rdar://145154078

Reviewed by Yusuke Suzuki.

This patch should strike a balance between convenience and performance.
Using a fixed array of callbacks is suitable because we always
know which systems are going to hook into libpas -- it's not like
someone will decide at runtime that they need to fall back on
the scavenger.

Re.: the additional mutex, it's possible to do without it -- we would
just need two atomic integers, "next_open_slot" and
"last_initialized_slot". However, this is a nontrivial amount of
complexity and much easier to mess up.

* Source/WTF/wtf/SequesteredAllocator.h:
* Source/WTF/wtf/SequesteredImmortalHeap.h:
(WTF::SequesteredImmortalHeap::installScavengerHook):
(WTF::SequesteredImmortalHeap::scavenge):
* Source/WTF/wtf/Threading.cpp:
(WTF::initialize):
* Source/bmalloc/libpas/libpas.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/src/libpas/pas_scavenger.c:
(ensure_data_instance):
(scavenger_thread_main):
(pas_scavenger_try_install_foreign_work_callback):
* Source/bmalloc/libpas/src/libpas/pas_scavenger.h:
* Source/bmalloc/libpas/src/test/ScavengerExternalWorkTests.cpp: Added.
(incrementCounter):
(testCallbacksAreCalledWhenExpected):
(addScavengerExternalWorkTests):
* Source/bmalloc/libpas/src/test/TestHarness.cpp:

Canonical link: https://commits.webkit.org/290978@main
